src/include/index/ivf_pq_group.h

```diff
@@ -0,0 +1,391 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * @todo This duplicates huge amounts of code from both ivf_flat and flat_pq,
+ * but there are enough differences that immediate reuse isn't obvious.  This
+ * should be refactored at some point.
+ */
+
+#ifndef TILEDB_IVF_PQ_GROUP_H
+#define TILEDB_IVF_PQ_GROUP_H
+
+#include "index/index_defs.h"
+#include "index/index_group.h"
+#include "index/ivf_pq_metadata.h"
+
+// flat_ivf_centroids, pq_ivf_centroids,
+// partitioned_pq_vectors, partitioned_pq_index, partitioned_pq_ids
+// cluster_centroids, distance_tables,
+
+[[maybe_unused]] static StorageFormat ivf_pq_storage_formats = {
+    {"0.3",
+     {
+         // @todo Should these be kept consistent with ivf_flat?
+         {"cluster_centroids_array_name", "pq_cluster_centroids"},
+         {"flat_ivf_centroids_array_name", "uncompressed_centroids"},
+         {"pq_ivf_centroids_array_name", "partition_centroids"},
+
+         {"ivf_index_array_name", "partition_indexes"},
+         {"ivf_ids_array_name", "shuffled_vector_ids"},
+
+         {"pq_ivf_vectors_array_name", "shuffled_vectors"},
+
+         {"distance_tables_array_name", "pq_symmetric_distance_tables"},
+     }}};
+
+template <class Index>
+class ivf_pq_group;
+
+template <class Index>
+struct metadata_type_selector<ivf_pq_group<Index>> {
+  using type = ivf_pq_metadata;
+};
+
+template <class Index>
+struct index_type_selector<ivf_pq_group<Index>> {
+  using type = Index;
+};
+
+template <class Index>
+class ivf_pq_group : public base_index_group<ivf_pq_group<Index>> {
+  using Base = base_index_group<ivf_pq_group>;
+
+  using Base::array_name_map_;
+  using Base::cached_ctx_;
+  using Base::group_uri_;
+  using Base::metadata_;
+  using Base::valid_array_names_;
+  using Base::valid_key_names_;
+  using Base::version_;
+
+  // @todo These should be defined in some common place and passed as parameters
+  static const int32_t default_domain{std::numeric_limits<int32_t>::max() - 1};
+  static const int32_t default_tile_extent{100'000};
+  static const int32_t tile_size_bytes{64 * 1024 * 1024};
+
+ public:
+  using index_group_metadata_type = ivf_pq_metadata;
+  // using index_type = Index;
```

Comment:
@lums658 could you help explain this? Is this a TODO to change to this later? Or can I remove it?